Sources/NautilusTelemetry/Tracing/URLTemplateMatcher.swift

```diff
@@ -0,0 +1,199 @@
+// Created by Jon Parise on 6/20/25.
+// Copyright © 2025 Airbnb Inc. All rights reserved.
+
+import Foundation
+
+// MARK: - URLTemplateMatcher
+
+/// Matches URLs against a collection of URL template patterns.
+///
+/// Template patterns describe a URL path and query string. They can use
+/// templated parameters like {} or {id} to match dynamic portions of the
+/// path and query parameter values (e.g. "/users/{id}?status={status}").
+///
+/// ## Template Syntax
+///
+/// - Path parameters: `/users/{userId}`
+/// - Query parameters: `?status={status}`
+/// - Literal query values: `?status=active`
+/// - Mixed: `/users/{id}?status={status}&verified=true`
+///
+/// ## Query Parameters
+///
+/// Query parameters can appear in any order to match. Any query parameters
+/// that aren't named in the template are ignored.
+///
+/// If a literal query parameter pair (e.g. "status=active") is included in
+/// the template, it must appear exactly that way in the URL for it to match.
+///
+/// ## Examples
+///
+/// ```swift
+/// let matcher = try URLTemplateMatcher(templates: [
+///     "/users/{id}",
+///     "/posts/{postId}/comments/{commentId}",
+///     "/search?q={}&status=active"
+/// ])
+///
+/// let url = URL(string: "https://api.example.com/users/123")!
+/// let matchedTemplate = matcher.match(url: url) // Returns "/users/{id}"
+/// ```
+///
+/// ```swift
+/// let span = tracer.startSpan(
+///     request: request,
+///     template: matcher.match(request.url)
+/// )
+/// ```
+public struct URLTemplateMatcher {
+
+	// MARK: Lifecycle
+
+	/// Creates a matcher with the given URL templates.
+	/// - Parameter templates: Array of URL template strings
+	/// - Throws: If any template contains invalid syntax
+	public init(templates: [String]) throws {
```

Comment:
I wonder if there's any value in doing `[StaticString]` here? I guess String/StaticString interop is a little obnoxious, but could help push the idea of low cardinality.